### PR TITLE
Add offset for tabled particle effects

### PR DESCRIPTION
**Pull request by Kestrellius**
This allows modders to specify a three-dimensional offset for a tabled particle effect, whereupon the position of the effect will be modified accordingly. This should work for effects spawned from objects, particles, vectors, and beams.

The offset is relative to the spawner's orientation. Unfortunately, this only works fully for objects. Other spawner types have a direction vector, but not a full orientation matrix, so in those cases the x and y axes are undefined. (In practice this seems to mean that they end up sort of pointed toward the global orientation to the extent that's possible.) This includes weapon hit effects, which are spawned from vectors, not the weapon object.

In future I would like to address this by adding an optional orientation field to SourceOrigin, which can then be populated with spawner orientation data where relevant -- such as the weapon's orientation for hit effects.

### CODE REVIEW

**Comment by Kestrellius** [comment]
(Note: I decided it would make sense to add particle parenting to this same branch, and also there's a bunch of complexity that needs to be worked out to do with turrets, so this is not actually ready yet; please nobody approve and merge it without warning me.)

**Comment by Kestrellius** [comment]
Okay. After a whole bunch of pain and suffering, parenting has been added for tabled particle effects, and it and offset now work correctly when effects are spawned from turrets as muzzle effects.

We also ended up fixing a variety of small bits of incorrect behavior in the process. I will attempt to document them.

**Comment by Kestrellius** [comment]
1. aibig.cpp, 1936: In ai_big_maybe_enter_strafe_mode, there was an assert that triggered if the weapon closest to the ship being examined didn't have a parent. At a certain point this started triggering regularly in the testing mission. I don't know why it suddenly started happening there, but in any case, we made it just return 0 (don't enter strafe mode) instead.
2. ship.cpp, 9290: Ship death effects were previously spawned in global space, ignoring the ship's orientation. (I strongly suspect, though I don't remember well enough to be sure, that this was responsible for thwarting an attempt I made once to produce a particular effect for FOTG.) With our addition of an explicit m_host_orientation field for effects spawned from vectors, ship death effects can now properly inherit the ship's orientation.
3. ship.cpp, 13790: Bit of complexity here. So, when spawning muzzle flashes for primary/secondary weapons, the ship_fire_[primary/secondary] functions previously fed the ship's fvec into shipfx_flash_create. (To be used, in theory, for the orientation of the offset of muzzle blobs.) This will obviously be incorrect in the case of weapon banks pointing in directions other than forward, which are admittedly rare. However! mflash_create, called from shipfx_flash_create, doesn't actually do anything with this vector; it figures out where the gun is pointing on its own and uses that. Muzzle effects, though (recently added by me), do use this direction vector. We arranged it so that muzzle effects have the ship object as their parent, and the gun's direction (relative to ship orientation) as their source orientation. This does mean that tabled position offsets for muzzle effects will be in the ship's orientation, which is unfortunate, but it's an edge case and for now it's difficult to fix. Anyway, after all that, the behavior of retail-style muzzle flashes doesn't change.
4. beam.cpp, 3962 and 3996: Beam piercing effects previously used SetOrientation (which affects SourceOrientation) to point the effect in the direction of the beam. Now (again using the explicit m_host_orientation field), the beam's direction vector is set as the effect's host orientation. This doesn't change behavior (apart from allowing tabled position offsets to work correctly in this case), but it's more conceptually consistent and should be more robust.
5. weapons.cpp, 5928: Similarly to the last example, weapon spawn effects (played when a weapon spawns a child weapon without dying) now have their host orientations set to the weapon's orientation, rather than having their SourceOrientations set to an orientation derived from the weapon's velocity. (Again, this shouldn't change existing behavior, since AFAIK weapons must always point in the direction of their velocity. I think primaries set to home used to be an exception before homing primaries were implemented, but they shouldn't be anymore.)
6. weapons.cpp, 7608, 7628, 7640, and 7684: Similar deal. Since these are hit effects, the weapon won't exist anymore by the time we're trying to play the effect, but we can use the host orientation field to store its orientation. This will allow positional offsets to work properly in all three dimensions in these cases.

File: code/particle/ParticleSource.h [hunk 1]
```diff
@@ -166,9 +193,22 @@ class SourceOrigin {
...
+ * 
+ * A source's SourceOrientation is distinct from its host orientation. The host orientation is either defined in
+ * SourceOrigin or gathered from the parent object, depending on host type. Host orientation is applied before position offset.
+ * SourceOrientation is applied after position offset.
+ * 
+ * 
+ * A source's SourceOrientation is distinct from its host orientation. The host orientation is either defined in
+ * SourceOrigin or gathered from the parent object, depending on host type. Host orientation is applied before position offset.
+ * SourceOrientation is applied after position offset.
+ *
```
[the remark this answers is not in the document]
Comment (Kestrellius) [reply code/particle/ParticleSource.h]: Taken care of!

**Review by wookieejedi — APPROVED** [review]
Given the code repetition in `ParticleSource` is a known issue and is planned on being fixed in subsequent large cleanup PRs related to particles in general, then this PR looks good to me. Though of course would be happy to have an additional set of eyes on it if anyone feels inclined! PR has also been tested and works as expected.

**Pull request merged**